# Let's review the ubike scraping code

In [10]:
import requests
import json
url  = 'http://data.taipei/youbike'
response = requests.get(url)
data = json.loads(response.text)
print("response code:", response.status_code)

response code: 200


# Prac: Use above code to get back the following data

In [38]:
url_AQX = "https://taqm.epa.gov.tw/taqm/aqs.ashx?lang=tw&act=aqi-epa&ts=1538961147679"
url_dcard = "https://www.dcard.tw/_api/forums/girl/posts?popular=true"
url_pchome = "http://ecshweb.pchome.com.tw/search/v3.3/all/results?q=X100F&page=1&sort=rnk/dc"
url_104 = "https://www.104.com.tw/jobs/search/list?ro=0&keyword=%E8%B3%87%E6%96%99%E5%88%86%E6%9E%90&area=6001001000&order=1&asc=0&kwop=7&page=2&mode=s&jobsource=n104bank1"

# Write a function to load json data

In [28]:
def get_web_json(url):
    response = requests.get(url)
    print("response code:", response.status_code)
    if not response.ok:
        return None
    data = json.loads(response.text)
    return data


In [34]:
data = get_web_json(url_591)
print(data)

response code: 404
None


In [39]:
response = requests.get(url_591)

In [57]:
response.ok

AttributeError: 'dict' object has no attribute 'ok'

## Use requests.json()

In [59]:
data = requests.get(url_dcard).json()
type(data)

list

In [62]:
data[0]

{'id': 229814982,
 'title': '跟20年的長髮掰掰👋',
 'excerpt': '小女 子 我從小時候有記憶以來就都是長頭髮，最短大概也是肩下5公分左右的中長髮吧🤔，每次看到身邊的人剪我都很想剪，但又很害怕自己剪起來很醜，加上母親又是長髮控，每次都說我短髮一定很醜，（哪有人這樣說自己女兒的？？？）',
 'anonymousSchool': True,
 'anonymousDepartment': True,
 'pinned': False,
 'forumId': 'f72e3b1d-3c9a-4fec-8a61-41c76cc317af',
 'replyId': None,
 'createdAt': '2018-10-09T03:45:46.180Z',
 'updatedAt': '2018-10-10T07:39:55.335Z',
 'commentCount': 69,
 'likeCount': 2060,
 'withNickname': False,
 'tags': [],
 'topics': ['髮型', '短髮'],
 'forumName': '女孩',
 'forumAlias': 'girl',
 'gender': 'F',
 'replyTitle': None,
 'supportedReactions': [],
 'reactions': [{'id': '286f599c-f86a-4932-82f0-f5a06f1eca03', 'count': 2060}],
 'hidden': False,
 'customStyle': None,
 'withImages': True,
 'withVideos': False,
 'media': [{'url': 'https://i.imgur.com/GtftTj1.jpg'},
  {'url': 'https://i.imgur.com/2Gypxwm.jpg'},
  {'url': 'https://i.imgur.com/CMlAhNf.jpg'},
  {'url': 'https://i.imgur.com/egqVQJN.jpg'},
  {'url': 'https://i.imgur.com/4z8d3tf.jp

# Prac: Problmatic url? See requests doc
* Try to get back the url https://rent.591.com.tw/home/search/rsList?is_new_list=1&type=1&kind=2&searchtype=1&region=1
* You will get an 404 status_code
* Solution: https://stackoverflow.com/questions/10606133/sending-user-agent-using-requests-library-in-python

## Parameters of a request
* Quick start of requests library http://docs.python-requests.org/en/master/user/quickstart/
```
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8
Accept-Encoding: gzip, deflate, br
Accept-Language: en-US,en;q=0.9,zh-TW;q=0.8,zh;q=0.7,zh-CN;q=0.6
Cache-Control: max-age=0
Connection: keep-alive
Cookie: webp=1; PHPSESSID=74131f35fb441ebacf685a6552f2629b; T591_TOKEN=74131f35fb441ebacf685a6552f2629b; urlJumpIp=1; urlJumpIpByTxt=%E5%8F%B0%E5%8C%97%E5%B8%82; 591_new_session=eyJpdiI6IjBqSkVXb04zTTkzV2FEeXZWUFlmc3c9PSIsInZhbHVlIjoiblZpOCtGcXpBS1gyeWJcL3dWM3N3c1BRbytFSEVIaXNud0VyMjA1ZmFBZVhXSHpLbnlmMW41YzQ0UU5GMXhMYTJoMGNOblJucGo1YXpzbDhPdUpBNEJnPT0iLCJtYWMiOiJlMDdjMjM2Y2Y3NDVjOTFkOWU5ZmMyNjY4YmZmNzRjOWY4MDBkZTY3ZTlhMzg2YzQ4ZGQ1ODI4MzI4NTM2ZWJiIn0%3D
Host: rent.591.com.tw
Upgrade-Insecure-Requests: 1
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_14_0) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/69.0.3497.100 Safari/537.36
```


In [44]:
url_591 = "https://rent.591.com.tw/home/search/rsList?is_new_list=1&type=1&kind=2&searchtype=1&region=1"
response = requests.get(url_591)
response.ok

False

In [51]:
headers = {
    'User-Agent': 'Chrome/69.0.3497.100'
} # headers are sent by dict form
response = requests.get(url_591, headers=headers).json()
# response.ok

In [63]:
type(response)
response.keys()
response["data"]["data"]

[{'id': 6784977,
  'user_id': 2169784,
  'address': '南海路2巷捷運站旁..',
  'type': '1',
  'post_id': 6784977,
  'regionid': 1,
  'sectionid': 1,
  'streetid': 25811,
  'room': 0,
  'area': 14,
  'price': '17,000',
  'storeprice': 0,
  'comment_total': 4,
  'comment_unread': 4,
  'comment_ltime': 1538901440,
  'hasimg': 1,
  'kind': 2,
  'shape': 3,
  'houseage': 0,
  'posttime': '3分鐘內',
  'updatetime': 1536627603,
  'refreshtime': 1539176402,
  'checkstatus': 1,
  'status': '',
  'closed': 0,
  'living': 'depart,advstore,market,night,park,school,hospital',
  'condition': 'tv,icebox,cold,washer,hotwater,four,broadband,landpost,trabus,balcony_1,bed,wardrobe,sofa,bookTable,chair',
  'isvip': 1,
  'mvip': 1,
  'is_combine': 1,
  'cover': 'https://hp1.591.com.tw/house/active/2018/09/11/153662657174722804_210x158.crop.jpg',
  'browsenum': 115,
  'browsenum_all': 4247,
  'floor2': 0,
  'floor': 3,
  'ltime': '2018-10-10 21:00:21',
  'cases_id': 0,
  'social_house': 0,
  'distance': 0,
  'search_nam

# Use crhome devtool to find json files
* Slide: https://docs.google.com/presentation/d/e/2PACX-1vRW84XoB5sFRT1Eg-GrK4smX23qoNkFffz_h8oRU4AIvJAgrrxBn8059_0UeHv_pFBks_Z37vNbLGai/pub?start=false&loop=false&delayms=3000

In [64]:
headers = {
    'User-Agent': 'Chrome/69.0.3497.100'
} # headers are sent by dict form

In [73]:
res = requests.get(url_591, headers=headers).json()
all_data = res["data"]["data"]
res.keys()
res["records"]

'2,414'

In [78]:
total_page = int(res["records"].replace(",", ""))
print(total_page)
url = "https://rent.591.com.tw/home/search/rsList?is_new_list=1&type=1&kind=2&searchtype=1&region=1&firstRow={}&totalRows={}".format(0*30, total_page)
url = "https://rent.591.com.tw/home/search/rsList?is_new_list=1&type=1&kind=2&searchtype=1&region=1&firstRow={}&totalRows={}".format(1*30, total_page)
url = "https://rent.591.com.tw/home/search/rsList?is_new_list=1&type=1&kind=2&searchtype=1&region=1&firstRow={}&totalRows={}".format(2*30, total_page)
url = "https://rent.591.com.tw/home/search/rsList?is_new_list=1&type=1&kind=2&searchtype=1&region=1&firstRow={}&totalRows={}".format(3*30, total_page)
res = requests.get(url_591, headers=headers).json()
res.keys()

2414


dict_keys(['status', 'data', 'records', 'is_recom', 'deal_recom', 'online_social_user'])

In [84]:
int(total_page/30)
all_data = []

In [87]:
all_data = []
for i in range(0, int(total_page/30)+1):
    url = "https://rent.591.com.tw/home/search/rsList?is_new_list=1&type=1&kind=2&searchtype=1&region=1&firstRow={}&totalRows={}".format(i*30, total_page)
    res = requests.get(url, headers=headers).json()
    all_data.extend(res["data"]["data"])
    print(i, len(all_data))

0 30
1 60
2 90
3 120
4 150
5 180
6 210
7 240
8 270
9 300
10 330
11 360
12 390
13 420
14 450
15 480
16 510
17 540
18 570
19 600
20 630
21 660
22 690
23 720
24 750
25 780
26 810
27 840
28 870
29 900
30 930
31 960
32 990
33 1020
34 1050
35 1080
36 1110
37 1140
38 1170
39 1200
40 1230
41 1260
42 1290
43 1320
44 1350
45 1380
46 1410
47 1440
48 1470
49 1500
50 1530
51 1560
52 1590
53 1620
54 1650
55 1680
56 1710
57 1740
58 1770
59 1800
60 1830
61 1860
62 1890
63 1920
64 1950
65 1980
66 2010
67 2040
68 2070
69 2100
70 2130
71 2160
72 2190
73 2220
74 2250
75 2280
76 2310
77 2340
78 2370
79 2400
80 2414


In [91]:
import pandas as pd
rent591_df = pd.DataFrame(all_data)
rent591_df.head()

addInfo  addition2  addition3  addition4 addr_number_name          address  \
0                  0          0          1             330號  信義路五段150巷象山旁新..   
1                  0          0          1              27號   塔悠路出租河堤景觀套房，..   
2                  0          0          1                   南京東路四段133巷近小巨..   
3                  0          0          1                    中華路一段English..   
4                  0          0          1                   仙岩路58巷科大世新獨立套..   

            address_img     address_img_title alley_name  allfloor    ...     \
0     象山旁新裝潢工業風小套房，雜費全包     象山旁新裝潢工業風小套房，雜費全包       150巷        13    ...      
1  出租河堤景觀套房，近饒河夜市，南港，內科  出租河堤景觀套房，近饒河夜市，南港，內科                   11    ...      
2  近小巨蛋.健康路.松山新城.育達.寧安街  近小巨蛋.健康路.松山新城.育達.寧安街       133巷         6    ...      
3  English新裝潢西門捷運電梯一房一廳  English新裝潢西門捷運電梯一房一廳                   12    ...      
4      科大世新獨立套房採光佳有陽台包網      科大世新獨立套房採光佳有陽台包網        58巷         6    ...      

   storeprice street_name  streetid  type unit  updatetime  user_id vipBorder  \
0           0       信義路五段     26234     1  元/月  1537977490  1702300  vipStyle   
1           0         塔悠路     26192     1  元/月  1537169020  1077174  vipStyle   
2           0      南京東路四段     26184     1  元/月  1539069855   413312  vipStyle   
3           0       中華路一段     26295     1  元/月  1538897991   266409  vipStyle   
4           0         仙岩路     25928     1  元/月  1538140234   129411  vipStyle   

   vipimg  vipstyle  
0             isvip  
1             isvip  
2             isvip  
3             isvip  
4             isvip  

[5 rows x 88 columns]

# Dump files for backup

## Dump one variable to json

In [92]:
import json
with open('rent_591.json', 'w') as outfile:
    json.dump(all_data, outfile)

## Dump multiple variables to pickle

In [95]:
import pickle
with open('rent591.pkl', 'wb') as f:  # Python 3: open(..., 'wb')
    pickle.dump([all_data, rent591_df], f)

## Load multiple variables back to objects

In [97]:
with open('rent591.pkl', "rb") as f:  # Python 3: open(..., 'rb')
    rent591_list, rent591_df2 = pickle.load(f)

In [99]:
len(rent591_list)

2414

In [101]:
rent591_df2.describe()

addition2  addition3    addition4     allfloor         area  \
count  2414.000000     2414.0  2414.000000  2414.000000  2414.000000   
mean      0.001243        0.0     0.032312     8.335128    10.194118   
std       0.035238        0.0     0.176863     5.474421     4.271879   
min       0.000000        0.0     0.000000     1.000000     3.000000   
25%       0.000000        0.0     0.000000     4.000000     7.000000   
50%       0.000000        0.0     0.000000     7.000000     9.500000   
75%       0.000000        0.0     0.000000    12.000000    13.000000   
max       1.000000        0.0     1.000000   121.000000    38.000000   

         browsenum  browsenum_all  checkstatus  closed  comment_ltime  \
count  2414.000000    2414.000000  2414.000000  2414.0   2.414000e+03   
mean     60.536040     850.626346     0.762635     0.0   7.610690e+08   
std      58.292908    1004.820779     0.937100     0.0   7.552827e+08   
min       0.000000       1.000000     0.000000     0.0   0.000000e+00   
25%      21.000000     229.000000     0.000000     0.0   0.000000e+00   
50%      44.000000     516.000000     1.000000     0.0   1.291179e+09   
75%      81.000000    1099.500000     1.000000     0.0   1.536105e+09   
max     574.000000   13206.000000     5.000000     0.0   1.539188e+09   

           ...        refreshtime  regionid         room    sectionid  \
count      ...       2.414000e+03    2414.0  2414.000000  2414.000000   
mean       ...       1.538912e+09       1.0     2.543496     5.552196   
std        ...       7.474037e+05       0.0    15.663882     3.047009   
min        ...       1.511853e+09       1.0     0.000000     1.000000   
25%        ...       1.538996e+09       1.0     0.000000     3.000000   
50%        ...       1.539136e+09       1.0     0.000000     5.000000   
75%        ...       1.539168e+09       1.0     0.000000     8.000000   
max        ...       1.539189e+09       1.0    99.000000    12.000000   

             shape  social_house  storeprice      streetid    updatetime  \
count  2414.000000   2414.000000      2414.0   2414.000000  2.414000e+03   
mean      1.714167      0.008285         0.0  26796.160315  1.538395e+09   
std       0.700120      0.090663         0.0   5449.111388  9.501974e+05   
min       1.000000      0.000000         0.0  25425.000000  1.511853e+09   
25%       1.000000      0.000000         0.0  25687.000000  1.538022e+09   
50%       2.000000      0.000000         0.0  25787.000000  1.538635e+09   
75%       2.000000      0.000000         0.0  26172.750000  1.538993e+09   
max       6.000000      1.000000         0.0  69947.000000  1.539189e+09   

            user_id  
count  2.414000e+03  
mean   1.043620e+06  
std    7.847464e+05  
min    1.258000e+03  
25%    3.074432e+05  
50%    9.255255e+05  
75%    1.708152e+06  
max    2.482651e+06  

[8 rows x 36 columns]